core/src/player.rs

```diff
@@ -628,14 +627,13 @@ impl Player {
                     ContextMenuCallback::Back => Self::back_root_movie(context),
                     ContextMenuCallback::Rewind => Self::rewind_root_movie(context),
                     ContextMenuCallback::Avm2 { item } => {
-                        // TODO: This should use the pointed display object (see comment on line 614)
-                        let root_dobj = context.stage.root_clip();
+                        let menu_item = *item;
 
-                        if let Some(root_dobj) = root_dobj {
-                            let menu_item = *item;
+                        if let Some(display_obj) = Player::get_context_menu_display_object(context)
```

Comment:
I don't think the object is re-picked once you click the option, right? I think we need to call the callback with the object that was originally right-clicked?

Reply:
It is indeed better if we can cache the picked object. But it is not certain that it is run on the right clicked object because it run's on the root if the right clicked object does not have a custom context menu defined. I can probably store the object in the GcRootData struct. But is that worth it?

Reply:
> because it run's on the root if the right clicked object does not have a custom context menu defined

No, this is normal - the "event" should propagate up the tree from the clicked DO until it encounters a DO with a menu (so at a top, the root).
Oh, I only now realized this isn't part of the PR. Then yeah, it needs to - otherwise the context menus will actually break in most cases, as usually when we right click, the cursor is on some specific DO that's a child of the menu owner; right?

Also you can experiment with FP events to observe this (we don't have these fields implemented on our side yet:
[menuSelect ](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/ContextMenu.html#event:menuSelect)
[menuItemSelect](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/ContextMenuItem.html#event:menuItemSelect)
And their fields contextMenuOwner, mouseTarget etc.

> It is indeed better if we can cache the picked object. 

It's not about caching, it's about correctness. when I right click on something and select an item but the mouse happens to hover over a different DO when clicking, then the code can run on a different object than expected.

> I can probably store the object in the GcRootData

We already do store related stuff in ContextMenuState.
ContextMenuCallback

Reply:
I can't find the displayObject stored in contextMenuState. I only seems to have a Vec of ContextMenuItem and a Vec of callbacks. Wich both don't seem to hold the clicked display object.

Reply:
> I can't find the displayObject stored in contextMenuState.

Yup, I meant to put the new field there instead of in GcRootData.

Reply:
Everything should have the correct behavior now. The function get_context_menu_display_object is now only called at a single place. I think that it makes the code more readable to have this functionality in a function. Or should I remove the funtion and just add the code where the function is being called?